### PR TITLE
[Prism] Terminate Job with CancelFn instead of panic

### LINKED ISSUE #31338: [Bug][Prism]: panic: nothing in progress and no refreshes with non zero pending elements
Java test `org.apache.beam.sdk.transforms.FlattenTest` fails on Prism with error `panic: nothing in progress and no refreshes with non zero pending elements`:

Steps to reproduce.

1) Run Prism server.

```
./gradlew :runners:prism:runServer
```

2) Execute `:runners:portability:java:ulrLoopbackValidatesRunnerTests` on Prism:

```
TEST=org.apache.beam.sdk.transforms.FlattenTest
./gradlew :runners:portability:java:ulrLoopbackValidatesRunnerTests -PjobEndpoint=localhost:8073 --tests="$TEST"
```

Parent tracker: #28187

### PR DESCRIPTION
**Pull request by damondouglas**
This PR closes #31338, refactoring `engine.ElementManager` `Bundles` method to receive a `context.CancelCauseFunc` so that the Job fails instead of panic. This allows us to associate failing validation tests with stuck jobs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

### CODE REVIEW

**Comment by damondouglas** [comment]
R: @lostluck

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go [hunk 1]
```diff
@@ -22,6 +22,7 @@ import (
 	"container/heap"
 	"context"
 	"fmt"
+	"github.com/apache/beam/sdks/v2/go/pkg/beam/internal/errors"
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go:25]: This import is in the wrong group. Please put it in the group with the other sdk imports, not the standard library imports.

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go [hunk 2]
```diff
@@ -290,9 +291,9 @@ func (rb RunBundle) LogValue() slog.Value {
 // Bundles is the core execution loop. It produces a sequences of bundles able to be executed.
 // The returned channel is closed when the context is canceled, or there are no pending elements
 // remaining.
-func (em *ElementManager) Bundles(ctx context.Context, nextBundID func() string) <-chan RunBundle {
-	runStageCh := make(chan RunBundle)
+func (em *ElementManager) Bundles(ctx context.Context, upstreamCancelFn context.CancelCauseFunc, nextBundID func() string) <-chan RunBundle {
 	ctx, cancelFn := context.WithCancelCause(ctx)
+	runStageCh := make(chan RunBundle)
```
Comment (lostluck) [inline sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go:296]: Move this back before the cancel line above to minimize the diff.

**Review by lostluck — APPROVED** [review]
Approved after suggested changes.

If you're feeling ambitious, we can write a test that a contrived/bad elementmanager configuration pipeline will call the passed in cancel function when it gets there. For obvious reasons we didn't have a test for that yet.

File: sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go (hunk 1 above)
Comment (damondouglas) [reply sdks/go/pkg/beam/runners/prism/internal/engine/elementmanager.go:25]: I'm so embarrassed. I'll fix it.

**Comment by damondouglas** [comment]
I wasn't successful creating a test case that simulates the behavior. I suspect as we fix other unsupported features, we might be able to reveal more what is going on. It's only one Java Test class and only one is it's methods. In the grand scheme of things, I'm ok moving on for now.

**Pull request merged**